### DISCUSSION TITLE
using saveExternalAuth

### DISCUSSION THREAD

**Discussion by mrpn** [pocketbase/pocketbase#3737] [Q&A] [answered]
I need advice on implementing `saveExternalAuth` with the Javascript extension in a specific scenario.

My users collection, sourced from the [Discord API](https://discord.com/developers/docs/resources/guild#list-guild-members), is structured like this:
| username | nick | avatar |
| --- | --- | --- |
| 132123122943 | user nick | avatarurl |
| 12348538425| user nick 2 | avatarurl |

So, when a user signs in via Discord oAuth, I want to link their auth to an existing record in my collection. Currently, a new user is created instead of linking to an existing record. I'm considering using `saveExternalAuth` within the `onRecordBeforeAuthWithOAuth2Request` hook, but I'm unsure if this is the best approach.

Any insights or suggestions would be greatly appreciated

**Comment by ganigeorgiev** [comment]
Hm, I'm not sure actually if you can use `$app.dao().saveExternalAuth()` because `*models.ExternalAuth` is not mapped (_I'll consider exposing it in future release_), but direct `INSERT` query into the `_externalAuths` table should work.

---

Usually the OAuth2 linking is done automatically based on the email field (if returned and there is a matching PocketBase user).

Alternatively, you can also "force" the linking by first authenticating the user with password and then send another request with the same SDK client for the OAuth2 auth request:

```js
await pb.collection("users").authWithPassword("user1", "1234567890");

// if this is the first time authenticating with discord, the OAuth2 account will be linked to the auth record from the previous call
await pb.collection("users").authWithOAuth2({ provider: "discord" });
```

The linking in this case works based on the `Authorization: TOKEN` header that it is send automatically with the request if `pb.authStore.token` is populated.

Both the manual and auto email linking works only if the OAuth2 user is not already linked to another PocketBase user.

**Reply by ganigeorgiev** [reply, marked as answer]
Another way to link a user during OAuth2 sign-up is to reasign the `e.record` field in the `onRecordBeforeAuthWithOAuth2Request` hook:

```js
onRecordBeforeAuthWithOAuth2Request((e) => {
  if (!e.record) {
    e.record = $app.dao().findAuthRecordByUsername("users", "something")
  }
}, "users")
```

**Reply by mrpn** [reply]
Such a neat solution and I can confirm it works perfectly. Thanks again for the quick reply
